runtime/oti/VMHelpers.hpp

```diff
@@ -1429,6 +1429,24 @@ class VM_VMHelpers
 			&& (ramClass->romClass->majorVersion >= 53);
 	}
 
+	/* Allocate an array instance
+	 * 
+	 * @returns the allocated array instance
+	 */
+	static VMINLINE j9object_t
+	inlineArrayAllocation(J9VMThread *currentThread, MM_ObjectAllocationAPI *objectAllocate, J9Class *arrayClass, U_32 size, bool initializeSlots = true, bool memoryBarrier = true, bool sizeCheck = true)
+	{
+		j9object_t instance = NULL;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		if (J9_IS_J9CLASS_FLATTENED(arrayClass)) {
+			instance = objectAllocate->inlineAllocateIndexableValueTypeObject(currentThread, arrayClass, size, initializeSlots, memoryBarrier, sizeCheck);
```

Comment:
This call does not appear to be initializing the array with the default value.

Reply:
Yes, inlineArrayAllocate calls are guarded by a `J9ClassContainsUnflattenedFlattenables` check if applicable. So inlineArrayAllocate is only called when it is known that the array elements can be initialized with zero.

Reply:
Is there a good reason this isn't just in the interpreter (then you would not need to pass the barrier object)?

Reply:
This can be moved to the interpreter.

@thomaslittletrainx ^^^

Reply:
Also, if the entire call is guarded with a VT test, then the ifdef in the middle of the code is just confusing (I'd ifdef the whole function when it's moved).